### ISSUE TITLE
heads up: gherkin3 

### ISSUE THREAD

**Issue by muhqu** [sittercity/gocumber#19]
Hi @cbandy, @ptrimble, @zombor,

awesome! …I'm very happy you've put my https://github.com/muhqu/go-gherkin to great use. :sparkles::tada: 

I'm not sure you're already aware of it, but cucumber is currently in the process of reimplementing it's gherkin parser and test runner. I've contributed a new version of the gherkin parser for golang, which is now compatible with all the grammar and language dialects. The main work gets done in a shared repository for all gherkin parser implementations (ruby, js, java, csharp, golang...) 
- https://github.com/cucumber/gherkin3

But there is also a go-get-able subtree split repository: 
- https://github.com/cucumber/gherkin-go

Maybe you folks are interested in contributing to a full-fledged go cucumber version. e.g. a tool that let users just type `go cucumber` to automatically find step definitions and run feature files...

/cc @aslakhellesoy, @nightlyone

**Comment by ppeble** [comment]
@muhqu Thanks for the information! I know that we have kicked around the idea of a `go cucumber` kind of tool internally. Initially we were hoping to just tack onto the go test framework but we have increasingly talked about a stand-alone tool like you proposed. It seems like a lot of what we want to do would be accomplished by going that route.

So if I understand correctly https://github.com/cucumber/gherkin-go is a replacement for go-gherkin? Are you planning on continuing to support the older project or should we have future plans to switch? How 'prod ready' is gherkin-go at this stage?

I have not talked specifically to @cbandy or @zombor about this but I'm sure we would be interested in helping out. Feel free to loop us in to discussions or issues that come up.

**Comment by muhqu** [comment]
@ptrimble [cucumber/gherkin-go](https://github.com/cucumber/gherkin-go) is the official implementation of a gherkin parser with support for [different dialects](https://godoc.org/github.com/cucumber/gherkin-go#GherkinDialectsBuildin) (e.g. languages) and  [muhqu/go-gherkin](https://github.com/muhqu/go-gherkin) is my own-grown gherkin parser which doesn't necessarily follows the 'standard'.

However, go-gherkin's tokenizer and grammar is somewhat more powerful then the new gherkin-go (parsing of line trailing comments, tags that are on the same line as scenarios...) but it lacks the support to change the tokenization rules while parsing which is required for `#language: ...` support.

I'm not sure yet, but I think I will not continue `go-gherkin` and keep focus on that more official and standards-compliant parser `cucumber/gherkin-go`...

Regarding how mature `cucumber/gherkin-go` is... well AKAIK no one is using it at the moment, but that's what I thought of `go-gherkin` before I stumbled upon your efforts, too. Please note that I implemented both parsers just for the fun of it and to get my hands on golang.

**Comment by ppeble** [comment]
@muhqu Fair enough! When I said 'prod ready' I guess I just meant that it was pretty much feature complete in the sense that it performs parsing correctly. We are using it for multiple projects internally and it's serving us well. The real complaints we have about `gocumber` are based around the things you already brought up: lack of a standalone tool such as `go cucumber` and other sugary things around that tool.

I'm of the mind to use the more compliant parser myself but I'll have a chat with @cbandy and @zombor. Either way, ping us when things come up or if there is work/reviews that you need. I'll personally take a look at the gherkin-go project in the near future and see what I can do to help.

**Comment by muhqu** [comment]
:+1:

**Comment by aslakhellesoy** [comment]
Hi guys - I've started on a reference implementation for Cucumber based on Gherkin3 in cucumber/gherkin3#37. Once this is a little more mature I'd love other Gherkin3-based Cucumber implementations to follow a similar design. That will make it so much easier to keep things more in sync.

For any questions about the design - please head over to cucumber/gherkin3#37

**Comment by ppeble** [comment]
FYI, I have left this open because we have started attempting to write a `gocumber` type command. It turns out that writing something like `go cucumber` is impossible since all of those other commands  (`go test`, `go run`) are baked into the language. As far as I am aware it is not possible to add a `go cucumber`, there is no mechanism for feeding custom commands like that. You would to entirely fork Go and add in your command, which is obviously a non-starter. :smile: 

In light of this we are starting by taking the existing `go test` source and seeing if we can modify it to give us cucumber-like behavior. Once we have something working we will be taking a look at the reference listed above and make sure we are matching expectations.

In in near term we will continue to use the go-gherkin parser unless whatever Gherkin3 stuff is in a solid position. We are taking care to ensure that a switch in the future would not be too difficult.

I wanted this note to go out so we could make sure a similar effort was not in progress. Let us know if you have questions or know of other efforts along these lines.

**Comment by ppeble** [comment]
And just after writing all of this we attended a Lightning talk at Gophercon by @kat-co regarding https://github.com/kat-co/gorkin. So we looked and realized that both https://github.com/lsegal/gucumber and https://github.com/DATA-DOG/godog both seem to be moving in the direction we want to go. AND it seems like the gherkin-3 parser is used by the godog implementation already!

We are looking into trying out the other tools on our own projects and consolidating. I'm closing out this issue in the meantime. Thanks!
